### ISSUE TITLE
Check dependencies for no_std crates

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust#38509]
It's possible to compile `no_std` crates with dependencies dependent on std. (e.g. [generic-array](https://github.com/fizyk20/generic-array) rev. #7f73b31) But as expected it's impossible to use such crate in environments without std. In my opinion when compiler builds `no_std` crate it also should check if `no_std` was enabled in all its dependencies, as otherwise we can't consider it independent of std.

**Comment by steveklabnik** [comment]
@rust-lang/lang , this would be technically-backwards-incompatible modification to `no_std`. What do you think?

**Comment by Amanieu** [comment]
Maybe just as a lint then? That would make it backward-compatible.

**Comment by steveklabnik** [comment]
New lints require an RFC.

**Comment by newpavlov** [comment]
For transitional period of course, but in my opinion this lint should state that in future it will become a hard error. I guess it will depend on how many crates (and versions of them) affected by this issue, if there is just a few of them then it's possible to skip lint stage altogether by contacting crate owners directly and arguing that it's better to yank affected versions, but if there is too many it will have to be just a lint, at least in mid-term. How hard it will be to get this number?

**Comment by bluss** [comment]
Seems like a great lint. Not sure it should ever be an error. Consider if you have a no_std crate with add-on std-using deps. It's more convenient if you can just continue to use `no_std` unconditionally.

**Comment by newpavlov** [comment]
I just afraid it will be a false guarantee. Personally when I see unconditional `#[no_std]` in the `lib.rs` I expect crate to be unconditionally independent over stdlib. (well, ideally I would like to get this infromation from `Cargo.toml` but it's a separate topic) In that context argument of "convenience" seems strange, especially in the language with goals of Rust. For me to have feature gated `std` dependent part of the code and to not disable `#[no_std]` for it is a bad practice, be it dependency, module or function. And compiler clearly forbids the later two. So why should it allow exception for dependencies?

Well if we'll go with the eternal lint approach, then this [part](https://doc.rust-lang.org/book/no-stdlib.html) of the book must be certainly changed:
>To do so, we tell Rust that we don’t want to use the standard library via an attribute: #![no_std].

**Comment by aturon** [comment]
I agree that this could make a good lint. cc @rust-lang/libs as well.

**Comment by nikomatsakis** [comment]
👍 to lint.

**Comment by nikomatsakis** [comment]
I don't think it should ever be an error. There are legitimate reasons to use `std` from a `no_std` crate. For example, if you have unit tests, then it is common for those tests to rely on standard even if the main code base does not. One could also imagine a crate that is `#[no_std]` much of the crate but offers cargo features that require `std`. However, in both of those cases I feel like one could just `#[allow]` the lint.

**Comment by newpavlov** [comment]
@nikomatsakis 
If you use something `std` dependent in your unit tests and will not disable `no_std` for tests you'll get compilation errors on `cargo test`. The second case will give a hard compilation error too. As I previously wrote the only exception (or more likely an overlook) is dependencies.

**Comment by nikomatsakis** [comment]
@newpavlov maybe I am not understanding, but I believe that one can just write `extern crate std` in your testing code, no?

**Comment by newpavlov** [comment]
@nikomatsakis 
You are right, I did not think of `extern crate std`.

Well, if it was not an overlook, but an intentional behaviour (with which I am not quite agree, but can understand) then lint and documentation update will do. I guess it's for @rust-lang/lang to decide.

**Comment by Ericson2314** [comment]
-1 The goal here would be better solved by a conjunction of https://github.com/rust-lang/rfcs/pull/1133 and a way to prohibit a crate from appearing in the build plan. For the latter one can use `[replace]` with a non-building crate as a stop-gap, but this is not intuitive.

**Comment by retep998** [comment]
Another case where having this be a hard error would be wrong is as follows. Crate A has an `std` feature that when disabled uses `#![no_std]`. Crate B is unconditionally `#![no_std]` and uses crate A without the `std` feature. Crate C also uses crate A but unlike B it uses the `std` feature. Crate D depends on both B and C, and due to C's requirements A is compiled with `std`, which means the A that B gets is using libstd. In this case the lint would fire on B because it is using `#![no_std]` while one of its dependencies isn't, and if it were a hard error this would be a really frustrating and pointless error that just serves to break everything.

**Comment by newpavlov** [comment]
@retep998 
If I understand correctly how Rust works with the crates, in this situation it will simply compile two different versions of A, one with enabled std feature and another without it. So B will link against one version and C against another.

**Comment by retep998** [comment]
@newpavlov That is not how cargo currently works. Right now if two crates depend on A with different feature sets, cargo will just use the union of those features and compile a single A for both crates to use.

**Comment by newpavlov** [comment]
@retep998 
In this case it can be solved by disabling this error/lint if top crate does not enable `#![no_std]`.

**Comment by newpavlov** [comment]
Looks like many crates in the ecosystem use code like this:
```rust
#![no_std]

#[cfg(feature = "std")]
extern crate std;
```

So even linting against it would be too disruptive. Thus I guess this issue can be closed.
